### PR TITLE
initial customizable grids

### PR DESCRIPTION
**Pull request by PhilipPartsch**
try to implement #612

### CODE REVIEW

File: sphinx_needs/layout.py [hunk 1]
```diff
@@ -354,8 +373,11 @@ def _func_replace(self, section_nodes):
                     func_def = line_element[0]
                     # Check if normal string was detected
                     if len(text) > 0 and len(func_def) == 0:
-                        node_line += nodes.Text(text, text)
-                        result = nodes.Text(text, text)
+                        if len(section_nodes) == 1:
+                            node_line = nodes.Text(text, text)
+                        else:
+                            node_line += nodes.Text(text, text)
+                        #result = nodes.Text(text, text)#dead code?
```
Comment (PhilipPartsch) [inline sphinx_needs/layout.py:380]: I belive it is dead code and can be removed.

File: sphinx_needs/layout.py [hunk 2]
```diff
@@ -297,6 +311,10 @@ def get_section(self, section: str) -> Optional[nodes.line_block]:
         # Needed for PDF/Latex output, where empty line_blocks raise exceptions during build
         if len(lines) == 0:
             return None
+        elif len(lines) == 1:
+            line_parsed = self._parse(lines[0])
+            line_ready = self._func_replace(line_parsed)
+            return line_ready
```
Comment (PhilipPartsch) [inline sphinx_needs/layout.py:317]: Here we now return in element with out lines_container (block). This leads to missing <span> elements in the ouput. Is this an issue?

File: sphinx_needs/layout.py [hunk 3]
```diff
@@ -407,14 +429,21 @@ def _func_replace(self, section_nodes):
...
+                                node_line += result
                     else:
                         raise SphinxNeedLayoutException(
                             f"Error during layout line parsing. This looks strange: {line_element}"
                         )
 
                 return_nodes.append(node_line)
-        return return_nodes
+
+        if len(section_nodes) == 1:
```
Comment (PhilipPartsch) [inline sphinx_needs/layout.py:443]: Ugly, but currently functional. I can refactor it, if we have an working solution.

File: sphinx_needs/layout.py [hunk 4]
```diff
@@ -1190,6 +1231,32 @@ def _grid_content(self, colwidths, side_left, side_right, footer):
...
+        for width in customer_format["widths"]:
+            node_colspec = nodes.colspec(colwidth = width)
+            node_tgroup += node_colspec
+
+        for r_k, r_v in customer_format["layout"].items():
+            row = nodes.row(classes=[r_k])
+            self.node_tbody += row
+
+            for c_k, c_v in r_v.items():
+                if "colspan" in c_v and "rowspan" in c_v:
```
Comment (PhilipPartsch) [inline sphinx_needs/layout.py:1247]: At first I had a for loop to add all elemnts, but I have to decrease these by one, so it is easier understable for the "user". Overall I have to think about, what is better:
- forward everything, without checking
- copy the know attributes

**Comment by PhilipPartsch** [comment]
Tests are failing, because they have to be adpated, becasue of breaking change.

File: sphinx_needs/layout.py [hunk 5]
```diff
@@ -134,6 +134,11 @@ def build_need(layout, node, app: Sphinx, style=None, fromdocname: Optional[str]
     # We need to replace the current need-node (containing content only) with our new table need node.
     node.parent.replace(node, new_need_node)
 
+# read configuration
+# ensure dict of array of arrays of tuble
+def getgridsfromconfig(app: Sphinx,):# -> dict:
```
Comment (danieleades) [inline sphinx_needs/layout.py:139]: this should use proper type annotations, not a type comment

File: sphinx_needs/layout.py [hunk 6]
```diff
@@ -407,14 +429,21 @@ def _func_replace(self, section_nodes):
...
 
                 return_nodes.append(node_line)
-        return return_nodes
+
+        if len(section_nodes) == 1:
+            return return_nodes[0]
+        else:
+            return return_nodes
 
     def _replace_place_holder(self, data):
```
Comment (danieleades) [inline sphinx_needs/layout.py:448]: type annotations?

File: sphinx_needs/layout.py [hunk 7]
```diff
@@ -956,6 +985,18 @@ def permalink(
             prefix=prefix,
         )
 
+    def content(self):
```
Comment (danieleades) [inline sphinx_needs/layout.py:988]: type annotations?

File: sphinx_needs/needs.py [hunk 8]
```diff
@@ -454,6 +455,129 @@ def visitor_dummy(*_args, **_kwargs) -> None:
     """
     pass
 
+#todo: test isattributeset
+def isattributeset(grid: {}, attributes: []) -> bool:
```
Comment (danieleades) [inline sphinx_needs/needs.py:459]:
```suggestion
def is_attribute_set(grid: {}, attributes: []) -> bool:
```

also, missing type annotations

**Review by danieleades — CHANGES_REQUESTED** [review]
use snake case for function names. A few missing type annotations as well

File: docs/conf.py [hunk 9]
```diff
@@ -285,6 +285,38 @@
 needs_default_layout = "clean"
 needs_default_style = None
 
+needs_grids = {
```
Comment (danieleades) [inline docs/conf.py:288]: these are really complex, stringly-typed dictionaries. Is it possible to replace any of the dicts used in this PR with classes?

**Comment by danwos** [comment]
I would like to close this PR, as nobody is working on it and it is out of any with master.
Any concerns?

**Pull request closed without merging**